User/Auth: Small Fixes & Validation Enhancements

fmSand opened the pull request:
## Small Fixes & Validation Enhancements

I know there are bigger changes and add-ons in the works, but I wanted to update a smaller chunk first.

### Changes Made

- Fix middleware role checks in isAdmin and isSelforAdm to use user.Role.name / user.role(sanitized )format. _(not that these are currently used)_
- Remove deprecated `displayName` field references . User model has a `fullName` field after prior update.
- Add empty body validation in `updateMe` controller ("At least one field must be provided for update")
- Enhance `updateUserSchema` with `bio` and `avatarUrl` fields + (optional tag after each field)
- Add strict validation with `noUnknown(true)` to prevent unknown fields
- Standardize error handling using `createError` and `successResponse` utilities
- `updateMe` no longer uses `sanitizedUser` directly in response, as `updatedUser` already returns `sanitizedUser`.

#### Testing (Optional)
- Combined existing tests with new validation tests
- 43 tests including schema validation, authentication, and CRUD operations

#### Additional Schema (Optional)
- Created `updatePasswordSchema` for future password update functionality

### Notes
- **Tests**: I've included more tests, but understand if you'd prefer to review/integrate them separately. Is mostly for my own testing while making changes.
- **Documentation**: I've started updating the API docs for user/auth endpoints but holding off on pushing those until we finalize more changes

 Would like to hear your thoughts. Happy to adjust or kill the PR!

fmSand committed:
feat: update user management validation, responses and test
- Add validation to updateUserSchema with bio, avatarUrl fields
- Add noUnknown(true)and strict() to prevent unknown fields in user updates
- Create updatePasswordSchema for secure password updates
- Add empty body validation in updateMe controller
- Fix middleware role checks to use consistent user.Role.name and  user.role format
- Replace manual error objects with createError utility for consistency
- Standardize success responses using successResponse helper
- Remove deprecated displayName field references from user model
- Remove santizeuser in updateMe as updated already returns santized
- Add temporary comprehensive test suite covering all user routes and validation

fmSand committed:
update test file

Anclagen commented on middleware/authentication.js:
```diff
@@ -87,7 +87,7 @@ async function isAdmin(req, res, next) {
     }
 
-    if (user.Role.role !== "admin") {
+    if (user.role !== "admin") {
```
Shouldn't this be?
```suggestion
    if (user.Role.name !== "admin") {
```

fmSand replied:
It's using getOneId:

``` const user = await userService.getOneId(decoded.id);```

and ```getOneId``` sanitizes the user:

``` role: user.Role?.name || "user",```

But could maybe be either or?

Anclagen reviewed (commented):
Looks fine, I'd add the updated docs before we merge.

More work can be added in the `feature-communications` where I have merged the email, password reset and verification code with the more recent dev updates. 

Once were going on the frontend more bugs and fixes will become apparent. We also need to add a flag to the user outputs to add on their projects, recently voted on projects etc..

fmSand committed:
update docs, and fix oopsi

fmSand commented:
@Anclagen I still haven't figured out why I'm suddenly getting 'password authentication failed for user "admin"' , so not been able to check that the docs looks ok. But no errors in terminal.

Anclagen commented:
What are you doing to get this error? The admin user in the seed is:
```json
  {
    "firstName": "Ada",
    "lastName": "Lovelace",
    "username": "admin_user",
    "email": "ada.admin@example.com",
    "password": "admin123",
    "bio": "Can override your code and your will to live. Handles bans with a smile.",
    "roleId": 3,
    "achievement": [1]
  },
  ```

There is some blocks in the guard that check roles and don't allow admin users that might need fixing but I was leaving those till we build an admin dashboard.

fmSand commented:
@Anclagen Sorry, I wrote that late last night and see now I should have added some context lol. I get the error in the terminal when trying to run the API. I got the same error a week or two ago, but then I hadn't updated the .env / credentials. So why I'm getting it again now, I don't know. I'll do some troubleshooting tonight.

fmSand committed:
docs: improve swagger deocumentation, and add user sanitization tests
- Add JWT error examples (missing, expired, invalid tokens)
- Include detailed validation error examplse for user endpoints
- Fix syntax errors in swagger JSDoc
- Add tests to verify owner-specific vs public field visibility
- Ensure API documentation matches actual error responses

fmSand commented:
@Anclagen Should be good now:)

Anclagen reviewed (changes requested):
Looks good. Before we merge I think something needs to be done about the get deleted user endpoint, at the moment it is returning this, with hashed passwords, perhaps we can just disable the end point for the moment.
```json
{
    "status": "success",
    "statusCode": 200,
    "data": [
        {
            "fullName": "John Doe",
            "id": "7646f62f-d3aa-4ca7-9df0-c8b3db9bad07",
            "firstName": "John",
            "lastName": "Doe",
            "username": "johndoe123",
            "email": "ada.admin@example.com",
            "hashedPassword": "f2677abbe4d30758d12592724d43c0103787bb8884bbdcd8c3c5ef0f9dfbb5d2",
            "salt": "e9053e7481e819b364f78f72230ef83f",
            "avatarUrl": "https://cdn.example.com/avatar/johndoe.jpg",
            "bio": "I'm a developer.",
            "roleId": 3,
            "createdAt": "2025-06-10T20:49:02.947Z",
            "updatedAt": "2025-06-13T08:28:06.865Z",
            "deletedAt": "2025-06-13T08:28:06.864Z"
        }
    ]
}
```

fmSand committed:
disable admin get deletes users

fmSand commented:
@Anclagen 
Added isAdmin handler, and serialize user with isAdmin to get-all-deleted-users, as well as commenting it out (controller, router, service) and removed the test and docs for it.

Anclagen reviewed (approved):
Looks good!
